### ISSUE TITLE
`performance_ex1` with triangles

### ISSUE THREAD

**Issue by aschaf** [mfem/mfem#4202]
Hi,

I was playing around with the performance miniapp `performance_ex1`, and in order to be able to run the program with the `-perf` and `-asm` option on _triangular_ meshes, I had to apply a small fix to `tbilinearform.hpp`. At
https://github.com/mfem/mfem/blob/37e03ab550e83cda2214e3b230e2aea5b24328c9/fem/tbilinearform.hpp#L505-L507
I replaced line 506 with
```
const int *dof_map_ = (dof_map) ? dof_map->GetData() : NULL;
```
which then allowed the program to run with `-perf -ams`. Second, I also tried using a `TPiecewiseConstCoefficient` and I found an error in 
https://github.com/mfem/mfem/blob/37e03ab550e83cda2214e3b230e2aea5b24328c9/fem/teltrans.hpp#L94-L106
where the increment in the inner loop increases the wrong variable. If these fixes are reasonable, I would open a pull request fixing those issues.

Best,
Andreas

**Comment by v-dobrev** [comment]
Your suggested fixes look good to me. Thanks @aschaf!

**Comment by aschaf** [comment]
I opened #4206 to resolve the issue. 

A related question: if I want to further speed up the assembly (eg. on my laptop) should I use the MPI parallel version to utilize all available cores, or would it make sense to somehow extend the TBilinearform to utilize eg. OpenMP  for the assembly loop?

**Comment by v-dobrev** [comment]
> A related question: if I want to further speed up the assembly (eg. on my laptop) should I use the MPI parallel version to utilize all available cores, or would it make sense to somehow extend the TBilinearform to utilize eg. OpenMP for the assembly loop?

The MPI-parallel version should be easier to try -- just look at how `miniapps/performance/ex1p.cpp` differes from `miniapps/performance/ex1.cpp`. Implementing an OpenMP version will require some effort and in the end may actually be slower.

**Comment by aschaf** [comment]
I noticed another issue I somewhat missed in the pull request. When turning on SIMD, it again crashes at line 103 in `teltrans.hpp`
https://github.com/mfem/mfem/blob/7c296d00d8a770e9f569e4db9d7e6d415902a886/fem/teltrans.hpp#L94-L106
specifically because at some point `el+j+i*vsize` is greater than the number of elements. A quick and dirty fix was just to replace line 103 with
```
attrib[i][j] = ((el+j+i*vsize) < fes.GetNE()) ? elements[el+j+i*vsize]->GetAttribute() : 1;
```

**Comment by v-dobrev** [comment]
@aschaf, you are right -- this is a bug. In the respective situation in `VectorExtract`, we stop copying values when we reach the last element, see https://github.com/mfem/mfem/blob/7c296d00d8a770e9f569e4db9d7e6d415902a886/fem/tfespace.hpp#L214

A better approach for here and for `VectorExtract` is probably to replicate the last element attribute or values, so that all SIMD entries have meaningful values.

**Comment by aschaf** [comment]
@v-dobrev Thank you for the explanation, I will open a PR once this is fixed.

**Comment by aschaf** [comment]
Would something like this be okay?
```
   void SetAttributes(int el, vint_t (&attrib)[NE]) const
   {
      const int vsize = sizeof(vint_t)/sizeof(attrib[0][0]);
      const int max_idx = fes.GetNE() - 1;
      for (int i = 0; i < NE; i++)
      {
         for (int j = 0; j < vsize; j++)
         {
            const int idx = std::min(el + j+i*vsize, max_idx);
            attrib[i][j] = elements[idx]->GetAttribute();
         }
      }
   }
```

And in `tfespace.hpp` use the code thats hidden by the preprocessor and add a similar `max_idx` construction?

**Comment by v-dobrev** [comment]
@aschaf, yes, this looks good to me.
